prover/src/key.rs

```diff
@@ -0,0 +1,149 @@
+use std::fs::File;
+use zstd::stream::{Decoder, Encoder};
+
+pub use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use nexus_nova::nova::pcd::compression::SNARK;
+
+use crate::error::*;
+use crate::pp::load_pp;
+use crate::srs::load_srs;
+use crate::types::*;
+use crate::LOG_TARGET;
+
+pub fn gen_key(pp: &ComPP, srs: &SRS) -> Result<SpartanKey, ProofError> {
+    let key = SNARK::setup(pp, srs)?;
+    Ok(key)
+}
+
+pub fn save_key(key: SpartanKey, file: &str) -> Result<(), ProofError> {
+    let f = File::create(file)?;
+    let mut enc = Encoder::new(&f, 0)?;
+    key.serialize_compressed(&mut enc)?;
+    enc.finish()?;
+    f.sync_all()?;
+    Ok(())
+}
+
+pub fn load_key(file: &str) -> Result<SpartanKey, ProofError> {
+    let f = File::open(file)?;
+    let mut dec = Decoder::new(&f)?;
+    let key = SpartanKey::deserialize_compressed_unchecked(&mut dec)?;
+    Ok(key)
+}
+
+pub fn gen_key_to_file(pp_file: &str, srs_file: &str, key_file: &str) -> Result<(), ProofError> {
+    tracing::info!(
+        target: LOG_TARGET,
+        path =?srs_file,
+        "Reading the SRS",
+    );
+    let srs = {
+        let mut term = nexus_tui::TerminalHandle::new();
+        let mut term_ctx = term.context("Loading").on_step(|_step| "SRS".into());
+        let _guard = term_ctx.display_step();
+
+        load_srs(srs_file)?
+    };
+
+    tracing::info!(
+        target: LOG_TARGET,
+        path =?srs_file,
+        "SRS found for a maximum of {} variables",
+        srs.max_num_vars
+    );
+
+    tracing::info!(
+        target: LOG_TARGET,
+        pp_file =?pp_file,
+        "Reading the Nova public parameters",
+    );
+
+    let pp: ComPP = {
+        let mut term = nexus_tui::TerminalHandle::new();
```

Comment:
ah I see. is it enough to drop the `_guard` to emit logs? so like this?
```
let mut term = nexus_tui::TerminalHandle::new();
let srs = {
    let mut term_ctx = ..."loading srs"..;
    let _guard = term_ctx.display_step();
    load_srs();
}
[emit logs]
let pp = {
    let mut term_ctx = ..."loading pp"..;
    let _guard = term_ctx.display_step();
    load_pp();
};
[emit more logs]
```

Reply:
Actually let's not spend more time on this, I just owe clear documentation and expected behavior of the whole TUI thing. Sorry you had to figure it out on your own.

Reply:
(regarding your question)
The idea is `tui` works with iterative tasks, including those consisting of a single step:
- `TerminalHandle` is global thing you need to create -- it spawns a thread.
- `TerminalContext` corresponds to "action" you're about to perform, creating it does nothing on its own.
- `Guard` is returned when you call `display_next_step`, while it's alive the timer is running (like `Computing ... 123ms`). When dropped, timer stops. If number of steps is reached, you automatically get `Finished` emitted into console, otherwise it waits for another `display_next_step`.

The main problem is that timer gets rendered on the same line, kind of carrying the cursor back to the beginning of the line. So if you emit something while `guard` is alive -- it will be overwritten.